### PR TITLE
remove invalid transactions on block finalization

### PR DESCRIPTION
**Pull request by dabasov**

### CODE REVIEW

File: code/go/0chain.net/chaincore/transaction/worker.go [hunk 1]
```diff
@@ -91,8 +91,45 @@ func RemoveFromPool(ctx context.Context, txns []datastore.Entity) {
...
+			current, ok := qe.(*Transaction)
+			if !ok {
+				logging.Logger.Error("generate block (invalid entity)", zap.Any("entity", qe))
+				return true
+			}
+
+			list := mappedTxns[txn.GetKey()]
+			for _, fromList := range list {
+				if fromList.GetKey() == current.GetKey() {
+					if current.Nonce <= fromList.Nonce {
```
Comment (peterlimg) [inline code/go/0chain.net/chaincore/transaction/worker.go]: The txn.GetKey() returns the transaction hash, so the nonce would always be the same as they are the same transaction.

**Review by peterlimg — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/chaincore/transaction/worker.go [hunk 2]
```diff
@@ -91,8 +91,45 @@ func RemoveFromPool(ctx context.Context, txns []datastore.Entity) {
...
+	var past []datastore.Entity
+	err := transactionEntityMetadata.GetStore().IterateCollection(cctx, transactionEntityMetadata, collectionName,
+		func(ctx context.Context, qe datastore.CollectionEntity) bool {
+			current, ok := qe.(*Transaction)
+			if !ok {
+				logging.Logger.Error("generate block (invalid entity)", zap.Any("entity", qe))
+				return true
+			}
+
+			list := mappedTxns[txn.GetKey()]
```
Comment (peterlimg) [inline code/go/0chain.net/chaincore/transaction/worker.go]: Seems the `txn` is an empty transaction, so I think the list is empty too.

File: code/go/0chain.net/chaincore/transaction/worker.go [hunk 3]
```diff
@@ -91,8 +91,45 @@ func RemoveFromPool(ctx context.Context, txns []datastore.Entity) {
...
 
+	logging.Logger.Debug("cleaning invalid transactions")
+	mappedTxns := make(map[string][]*Transaction)
+	for _, e := range txns {
+		blockTx, ok := e.(*Transaction)
+		if !ok {
+			logging.Logger.Error("generate block (invalid entity)", zap.Any("entity", e))
+			continue
+		}
+		mappedTxns[blockTx.GetKey()] = append(mappedTxns[blockTx.GetKey()], blockTx)
```
Comment (peterlimg) [inline code/go/0chain.net/chaincore/transaction/worker.go]: What's the mappedTxns key, seems it's txn hash, but perhaps should be client id.

**Review by peterlimg — DISMISSED** [review]

**Review by boddumanohar — APPROVED** [review]
LGTM

**Pull request merged**